workflow/controller/operator.go

```diff
@@ -1208,9 +1208,9 @@ func getExitCode(pod *apiv1.Pod) *int32 {
 	return nil
 }
 
-func (woc *wfOperationCtx) cleanUpPod(pod *apiv1.Pod) {
+func (woc *wfOperationCtx) cleanUpPod(pod *apiv1.Pod, tmpl wfv1.Template) {
 	for _, c := range pod.Status.ContainerStatuses {
-		if (c.Name == common.WaitContainerName || c.Name == common.MainContainerName) && c.State.Terminated == nil {
+		if (c.Name == common.WaitContainerName || tmpl.IsMainContainerName(c.Name)) && c.State.Terminated == nil {
```

Comment:
this logic here is confusing to me, I think we want to terminate the containers, if the main and wait containers have exited, is there a better way to formalize that?

Reply:
Agree it's confusing. Also currently we are calling `woc.queuePodForCleanup` multiple times which is unnecessary. I've refactored a bit.